layer/executables/packager.py

```diff
@@ -0,0 +1,154 @@
+import inspect
+import os
+import pickle  # nosec
+import shutil  # nosec
+import sys
+import tempfile
+import zipapp
+from pathlib import Path
+from typing import Any, Callable, List, Optional
+from uuid import uuid4
+
+import cloudpickle  # type: ignore
+
+
+def package_function(
+    function: Callable[..., Any],
+    resources: Optional[List[Path]] = None,
+    pip_dependencies: Optional[List[str]] = None,
+    output_dir: Optional[Path] = None,
+) -> Path:
+    """Packages layer function as a Python executable."""
+
+    with tempfile.TemporaryDirectory() as source_dir:
+        source = Path(source_dir)
+
+        requirements_path = source / "requirements.txt"
+        with open(requirements_path, mode="w", encoding="utf8") as requirements:
+            requirements.write(
+                "\n".join(_runtime_requirements + (pip_dependencies or []))
+            )
+
+        main_path = source / "__main__.py"
+        with open(main_path, mode="w", encoding="utf8") as main:
+            main.write(_loader_source())
+
+        if resources:
+            _package_resources(source, resources)
+
+        function_path = source / "function"
+        with open(function_path, mode="wb") as function_:
+            # register to pickle by value to ensure unpickling works anywhere
+            cloudpickle.register_pickle_by_value(sys.modules[function.__module__])
+            cloudpickle.dump(function, function_, protocol=pickle.DEFAULT_PROTOCOL)
+
+        # executable name is the function name (and not all callables always have a name)
+        executable_name = getattr(function, "__name__", f"callable-{uuid4()}")
+
+        target = (output_dir or Path(".")) / executable_name
+
+        # create the executable
+        zipapp.create_archive(
+            source, target, interpreter="/usr/bin/env python", compressed=True
+        )
+
+        # ensure the archive is executable
+        target.chmod(0o744)
+
+        return target
+
+
+def _package_resources(source: Path, resources: List[Path]) -> None:
+    resource_root = source / "resources"
+    resource_root.mkdir(parents=True, exist_ok=True)
+
+    def _copy_resource_file(resource_file: Path) -> None:
+        target_path = resource_root / resource_file
+        target_path.parent.mkdir(parents=True, exist_ok=True)
+        shutil.copyfile(resource_file, target_path)
+
+    def _walk_dir(dir: Path) -> None:
+        for root, dirs, files in os.walk(dir):
+            root_path = Path(root)
+            for file in files:
+                _copy_resource_file(root_path / file)
+            for dir_ in dirs:
+                _walk_dir(root_path / dir_)
+
+    for resource in resources:
+        if resource.is_file():
+            _copy_resource_file(resource)
+        elif resource.is_dir():
+            _walk_dir(resource)
+
+
+# depenedencies required by the runtime itself
+_runtime_requirements = ["cloudpickle==2.1.0"]
```

Comment:
nit: move this above, make it all caps, fix the typo in the comment

Reply:
What if `pip_requirements` also have `cloudpickle`?

Reply:
> What if `pip_requirements` also have `cloudpickle`?

It will fail. I've added a test which is skipped for now and created a ticket to address, as it is dependency resolution issue, not packaging.